implementations/rust/ockam/ockam_api/src/multiaddr_resolver/remote_resolver.rs

```diff
@@ -0,0 +1,257 @@
+                Ip4::CODE => {
+                    if transport_hop_resolved {
+                        return Err(multiple_transport_hops_error(ma));
+                    }
+
+                    let ip4 = p.cast::<Ip4>().ok_or_else(|| invalid_multiaddr_error(ma))?;
+
+                    (*ip4).to_string()
+                }
+                Ip6::CODE => {
+                    if transport_hop_resolved {
+                        return Err(multiple_transport_hops_error(ma));
+                    }
+
+                    let ip6 = p.cast::<Ip6>().ok_or_else(|| invalid_multiaddr_error(ma))?;
+
+                    (*ip6).to_string()
+                }
+                DnsAddr::CODE => {
+                    if transport_hop_resolved {
+                        return Err(multiple_transport_hops_error(ma));
+                    }
+
+                    let host = p
+                        .cast::<DnsAddr>()
+                        .ok_or_else(|| invalid_multiaddr_error(ma))?;
+
+                    (*host).to_string()
+                }
+                Worker::CODE => {
+                    let local = p
+                        .cast::<Worker>()
+                        .ok_or_else(|| invalid_multiaddr_error(ma))?;
+                    rb = rb.append(Address::new_with_string(LOCAL, &*local));
+                    continue;
+                }
+                Service::CODE => {
+                    let local = p
+                        .cast::<Service>()
+                        .ok_or_else(|| invalid_multiaddr_error(ma))?;
+                    rb = rb.append(Address::new_with_string(LOCAL, &*local));
+                    continue;
+                }
+                Secure::CODE => {
+                    let local = p
+                        .cast::<Secure>()
+                        .ok_or_else(|| invalid_multiaddr_error(ma))?;
+                    rb = rb.append(Address::new_with_string(LOCAL, &*local));
+                    continue;
+                }
+                _ => {
+                    return Err(Error::new(
+                        Origin::Api,
+                        Kind::Misuse,
+                        format!("Unsupported multiaddr protocol: {}", ma),
+                    ));
+                }
+            };
+
+            let connection = self.connect(ma, &mut it, &peer).await?;
+            transport_hop_resolved = true;
+            flow_control_id = Some(connection.flow_control_id().clone());
+            rb = rb.append(connection.sender_address().clone());
+            connection_res = Some(connection);
+        }
+
+        Ok(RemoteMultiaddrResolverResult {
+            flow_control_id,
+            connection: connection_res,
+            route: rb.into(),
+        })
+    }
+
+    async fn connect_tcp(&self, ma: &MultiAddr, peer: String) -> Result<TcpConnection> {
+        let tcp = self
+            .tcp
+            .as_ref()
+            .ok_or_else(|| invalid_multiaddr_error(ma))?;
```

Comment:
This code path is unreachable, since by the time this function is called we already checked that we have tcp transport, but I didn't want to use unwrap to be safe. I'll add more details to the comment. I don't think we should add a specific error for the unreachable path though